UI Touchup

codetheweb opened the pull request:
Highlights:
- Added [linter](https://www.npmjs.com/package/xo) (can swap it out if you want, I like it because it's strict)
- Converted class-based components to functional components
- Dark mode (changes based on system setting)
- A few UI tweaks
- Removed Thunk dependency for Redux, simplifies code

Preview:
![Screen Shot 2020-04-23 at 8 19 46 PM](https://user-images.githubusercontent.com/7410405/80164907-c800a080-859f-11ea-8af6-4522d5d3e8bb.png)

Some ideas for further development:
- ~~Mobile is garbage. Gotta figure out a good way to display tables.~~ After increasing responsiveness, it's not actually that bad. The table works fine with horizontal scrolling.
- ~~Increase performance when searching. @Blu3spirits suggested using a debouncer on user input. I changed the default table length to 10 rows and that seems to help as well, although obviously not an ideal solution.~~
- ~~Make API endpoint URL an env var so we can set it at build time.~~
- In line with the above, make some kind of script that creates a static build and links it with Django. **Edit**: maybe? I guess they recommend using a different server for static files.
- In the screenshot, you may see I added a 'last updated at...' value. I think this would be helpful for people who use the site but may think it was abandoned and doesn't have up-to-date information.

codetheweb committed:
Update UI

codetheweb committed:
Remove custom Babel config

codetheweb committed:
Fix table and loader

codetheweb committed:
Remove styled-components dependency

codetheweb committed:
Move XO to dev dependencies

codetheweb committed:
Fix Github Action linter

Blu3spirits commented:
Going to have to put this on the backburner for now as exams are this week.
Planning to get back to this Wed. Apr 29

codetheweb committed:
Speed up searching

codetheweb commented:
Used a combination of debouncing and memoization to significantly improve responsiveness.  Feels much snappier, especially on mobile.

codetheweb committed:
Add env var to control API endpoint

codetheweb committed:
Make default page size 50 rows

Blu3spirits commented:
> In line with the above, make some kind of script that creates a static build and links it with Django. Edit: maybe? I guess they recommend using a different server for static files.

**typically** it is recommended to do this. However, using a module such as `whitenoise` is supported and "okay" to do. I'm going to handle doing this one as it's on the Django side and would help solve a deployment hang-up I've been having.

Blu3spirits commented on webapp/src/ui/components/courses-list.js:
```diff
@@ -22,7 +22,7 @@ const columns = [
 const CoursesList = ({data}) => {
```
Not sure if this is a git rendering bug or not but all of these are 4 spaces on github's UI and then 2 in the code. This could also be a fault of my RC files automatically taking control. Our existing codebase uses 2 spaces and I see no reason to change from 2 to 4

codetheweb replied:
They're tabs, not spaces - not sure why they're rendering as 4 spaces.
I'll change the linter to use spaces instead.

codetheweb replied:
Fixed by 762accb. Also updated ESLint config for editor integration.

Blu3spirits commented on webapp/src/index.js:
```diff
@@ -1,21 +1,44 @@
+		() =>
+			createMuiTheme({
+				palette: {
+					type: prefersDarkMode ? 'dark' : 'light'
```
After downloading your PR I wasn't able to actually get the dark theme to work. I looked around and found it's actually the browser settings that control this, correct?

Can we make dark mode the default and then add a toggle for the people that want light mode?

codetheweb replied:
Correct, it's currently controlled by the OS setting (works on macOS and Windows in Chrome and Firefox, not sure about Linux desktop environments).

How about we make the default the system setting, and then the user can change it if they want with a toggle?

Blu3spirits replied:
Works for me, as long as there's some way to change the theme alongside OS themes.

codetheweb replied:
Check out 30fccb5. Works well on desktop, but kinda awkward on mobile:
![Screen Shot 2020-04-29 at 9 25 12 PM](https://user-images.githubusercontent.com/7410405/80665812-16aab080-8a60-11ea-9053-156cf7dee334.png)

codetheweb replied:
I'm gonna mark this as resolved for now. An easy way to fix mobile (and desktop for that matter) positioning of the theme button would be to add a navbar, but we don't need that yet.

Blu3spirits commented on webapp/src/store/actions/courses.js:
```diff
@@ -0,0 +1,16 @@
+export const GET_COURSES_ERROR = 'GET_COURSES_ERROR';
+
+export const getCourses = () => createAction({
+	endpoint: `${ENDPOINT}/api/courses/?limit=11000`,
```
Awesome idea with the config! This was something I had to change in my deployment process

codetheweb replied:
👍

Blu3spirits commented on webapp/src/ui/courses.js:
```diff
@@ -1,100 +1,93 @@
+	const dispatch = useDispatch();
+
+	// Update courses on load
+	useEffect(() => {
```
If we want to use this pattern, we should consider re-implementing thunk and start using selectors as well
See:
https://github.com/reduxjs/redux/issues/1171
https://github.com/reduxjs/redux-thunk#why-do-i-need-this

Also consider maybe looking into React's Lazy loading
https://reactjs.org/docs/code-splitting.html#reactlazy

It sort of does what we're doing already with the "if no data display a loading screen" logic but makes it require less boiler plate.

codetheweb replied:
I read through a couple different posts about redux thunk patterns ([this was really helpful](https://daveceddia.com/what-is-a-thunk/)), and I now have a much better understand of it than when I removed it. 😛 As far as I can tell, it doesn't make sense to use `redux-thunk` *until* we want to write our own thunks, as asynchronous network requests are currently handled by `redux-api-middleware`.

Where would you want to use selectors that we're not already? The only place we pull data from the store for display currently is here:
```js
const courses = useSelector(state => state.courses);
```

I'm not sure if lazy loading does what you think, AFAIK it's made more for code splitting (for example, don't load JS for page X until user hovers over link to page X).

Blu3spirits replied:
More what I was talking about with the selectors is breaking them apart from the main components, and placing them in the directory tree elsewhere. More on that point was there any specific reason the directory tree is now:
```
.
├── config.js
├── index.js
├── store
│   ├── actions
│   │   └── courses.js
│   ├── configure-store.js
│   └── reducers
│       ├── courses.js
│       └── index.js
└── ui
    ├── components
    │   └── courses-list.js
    ├── courses.js
    └── page.js
```
?

React Lazy was just something cool I was looking at and thinking about switching to in other projects rather than custom loading logic.

You are right in that we don't actually need to use redux-thunk until we want to write our own custom thunks. Thank you for the resource.

codetheweb replied:
Got it.

No, I just used the existing structure, and tried to organize Redux a bit by putting it all under `store/`. I'm open to other ways of structuring if you have ideas.  I usually do something like
- pages
	- _app.js (container)
 	- index.js (home page)
	- etc
- components
- lib

Blu3spirits replied:
In the past, I've used:

- src/
  - _some app.js
    - components/
    - App.js (Name doesn't really matter)
    - reducers.js
    - selectors.js
    - actions.js
  - _some app2.js
  - components/
etc.

Though for a small project it might not be necessary to break it up as much. It doesn't really make too much sense to stuff everything redux under `store` in my opinion.

codetheweb replied:
Is this more what you were thinking? 
db591d7

Blu3spirits commented on webapp/src/ui/page.js:
```diff
@@ -1,56 +1,39 @@
-  }
-}
+			<Typography variant="subtitle1" align="right">
+				Last updated <b>6 hours ago</b>
```
With the light theme variant this does not display

codetheweb replied:
Hmm, it displays for me:
![image](https://user-images.githubusercontent.com/7410405/80650458-a5580700-8a39-11ea-834b-8e6b80f84204.png)

As it's just a mockup of a feature that's not implemented on the backend yet, do you want to remove it for now (and make a new issue if you want it added at some point)?

Blu3spirits replied:
Weird, I could have sworn I didn't see it. Yeah, for now, let's disable this and then actually get it implemented in the backend.

Blu3spirits reviewed (commented):
I really like the dark theme idea you've put forward. Definitely a lot easier on the eyes and with debounced input the UX is vastly improved.

codetheweb committed:
Use spaces instead of tabs

codetheweb committed:
Remove "updated at" text mockup

Blu3spirits commented on webapp/.env.example:
```diff
@@ -0,0 +1 @@
+REACT_APP_ENDPOINT=http://localhost:8000
```
Without a valid configuration for the endpoint (missing port), the app will throw `Unhandled Rejection (TypeError): action.payload is undefined`

We should catch this behavior and display that something is wrong.

Blu3spirits replied:
Resolving, this comment is in the wrong location

Blu3spirits commented on webapp/src/config.js:
```diff
@@ -0,0 +1,3 @@
+const ENDPOINT = process.env.REACT_APP_ENDPOINT;
+
```
Without a valid configuration for the endpoint (missing port), the app will throw Unhandled Rejection (TypeError): action.payload is undefined

We should catch this behavior and display that something is wrong.

codetheweb replied:
Where would you want to add error checking code, and would it output to the console or to the browser?

IMO, there's not much of a point to error checking env vars since it's easy for the developer to test if it working and users will never be affected.

We could display an error if any `redux-api-middleware` actions fail.

Blu3spirits replied:
Kind of just thinking console, I was just surprised it didn't trigger GET_COURSES_ERROR from the RSAA.

codetheweb replied:
Were you thinking something like f856a6c?

The error reducer correctly triggered for me when I set `REACT_APP_ENDPOINT=http://localhost`. 🤷

Blu3spirits commented on webapp/src/store/actions/courses.js:
```diff
@@ -0,0 +1,16 @@
+export const GET_COURSES_ERROR = 'GET_COURSES_ERROR';
+
+export const getCourses = () => createAction({
+  endpoint: `${ENDPOINT}/api/courses/?limit=11000`,
```
I wonder if we can come up with a way to get around this hardcoded limit... I've been trying to come up with solutions for a while but the only things I can think of are slow or have a host of other issues, for example, hitting the API once to see how many results there are, then making the real request to get all of them. We should chat at some point and try to brainstorm ideas

codetheweb replied:
We could just replace 11,000 with an absurdly large number. 😛 
Otherwise we could add a method to the API that returns all results without pagination.

More generally, I was thinking that if we modified the backed to just be a script that spewed out a 2 MB JSON file, the site could be completely static - but that limits us in the future from adding features that actually require a backend store.

Blu3spirits replied:
We could just disable pagination on the API end, but that would make the browsable API (currently able to be exposed in planned "prod") basically unusable in browsers, which sort of defeats the point of making it available. I created issue #25 for this since it's not really needed for the scope of this PR

Kiro47 replied:
The easiest and least "guessy" solution we could have here is just to create a new `meta` endpoint in the api that contains some basic information about the data set.  Perhaps some info such as `last_updated`, `num_transfers`, etc.

That'll probably be something with it's own PR and feature improvement if we were to go that route, however.

codetheweb replied:
Moving to #25.

codetheweb committed:
Add theme selector

codetheweb committed:
Set default theme to OS preference

Kiro47 committed:
Refactor banwebScrape into a more modular style

This modular style will be much more useful as this begins to expand
more in the near future.

Kiro47 committed:
Add IDEs and venv to gitignore

Kiro47 committed:
Convert data_gathering and data_scraping to class based

Kiro47 committed:
Add logging framework to make consistently useful logs

Not many log messages currently, but that will quickly change
considering the complexity increase that will soon be happening.

codetheweb commented on webapp/src/ui/courses.js:
```diff
@@ -1,100 +1,93 @@
-const mapDispatchToProps = {
-  getCourses
-}
+      <Backdrop open={courses.loading} className={classes.backdrop}>
```
@Blu3spirits what do you think about using [skeletons](https://material-ui.com/api/skeleton/) here instead of the backdrop loader?

Blu3spirits replied:
Try it out and see. I like the idea of it but don't know how the implementation would look off the top of my head

codetheweb replied:
ae7a02b:

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/7410405/80854593-f5cc9180-8bfe-11ea-894f-8acfa7736fda.gif)

(Search field is disabled until load is complete.)

I like it, although the skeletons do cause moderate CPU usage in both Firefox and Chrome.

Blu3spirits replied:
For desktop, this would be fine, but mobile users would have it rough. I think we should go back to how it was.

Tested this on my cell and confirmed it's not a very good experience and the old way was better IMO

codetheweb replied:
Reverted.

codetheweb committed:
Remove debug console.log statement

codetheweb committed:
Fix inital value for theme

codetheweb committed:
Add error message for network error

codetheweb committed:
Restructure Redux

codetheweb committed:
Remove wrong comment

Blu3spirits reviewed (commented):
Good work!

I'm going to attempt to deploy this real quick before I give my approval

codetheweb committed:
Use skeleton loaders

Blu3spirits commented on webapp/.env.example:
```diff
@@ -0,0 +1 @@
+REACT_APP_ENDPOINT=http://localhost:8000
```
During deployment, this actually caused a bit of an issue.

Let's keep the ENDPOINT idea but instead apply it in package.json
under the proxy key/value pair

Here's what I had to hack together to get it to work:
```
+  "proxy": "https://localhost:8000/",
   "dependencies": {
     "@material-ui/core": "^4.9.8",
     "@material-ui/icons": "^4.9.1",
diff --git a/webapp/src/actions.js b/webapp/src/actions.js
index f9b89da..ac59b37 100644
--- a/webapp/src/actions.js
+++ b/webapp/src/actions.js
@@ -7,7 +7,7 @@ import {
 } from './action-types';

 export const getCourses = () => createAction({
-  endpoint: `${ENDPOINT}/api/courses/?limit=11000`,
+  endpoint: 'api/courses/?limit=11000',
```

Without this, [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) will fail

Then in the actions file just do endpoint: 'api/courses/'

codetheweb replied:
You're only supposed to use that field [during development](https://create-react-app.dev/docs/proxying-api-requests-in-development/), and I don't really want to commit env specific info to Git.

What were the issues during deployment?

Blu3spirits replied:
The issue without this is that nginx ingress even with allowing CORS still fails with this return code:
```
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://localhost:8000/api/courses/?limit=11000. (Reason: CORS request did not succeed).
```

I understand it's not exactly ideal for a production environment. But currently, it's the only way I have found that allows all of this to be run in the same docker environment.

Articles/technologies I have tried to get this to work:
https://imti.co/kubernetes-ingress-nginx-cors/
https://pypi.org/project/django-cors-headers/
https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#enable-cors

codetheweb replied:
Resolved in DM, if changes need to be made that'll be in another PR.

codetheweb committed:
Revert "Use skeleton loaders"

This reverts commit ae7a02b1e499adcca7ee1a93972d0439ec0154ec.

Kiro47 committed:
Removed commented out code

Kiro47 commented:
I did find one minor bug, but it can be moved to another issue if you feel like dealing with it later.

Recreation:

1. Do normal development setup
2. Go to web page
3. Go up a few pages to where you're in a higher results list.  (In example, I went to "Rows per Page=200" then went to the 4th page "Entries 601-800".
4. Search for something with less results than the lowest entry.  (An example being "Dynamics" with 72 results).
5.  On load it'll still have you on the page with 601-800 entries, and to get to your entries you'll just have to keep going back pages unless you're on actual values.

Proposed Solution:
On search query reset the page to page 1 with entries "1 to `Max_Rows`".

codetheweb commented:
I can fix that tonight.

Looks like I also have to start signing commits.

codetheweb committed:
Reset current page upon search change

codetheweb commented:
Fixed @Kiro47.  Edit: still working on signing old commits.

Blu3spirits commented:
If you wanted you could probably just rebase if you want. It's not exactly "best practice" but it will get everything under one signed commit

Kiro47 commented:
Looks like you missed the topmost commit when resigning.  To make it to resign here's a cheap little snippet:
```bash
git filter-branch -f --commit-filter 'git commit-tree -S "$@";' 6ac2b59359de934bc10252a76e4c39b2d9a57377^1..HEAD
```
It'll resign all the commits from that hash (`6ac2b59359de934bc10252a76e4c39b2d9a57377`) and on.  Since each commit hash is based on the old one, you unfortunately have to resign all the commits after modifying even one, this makes it a bit easier.
Also, make sure you keep that `^1`, otherwise it will gpg resign everything *after* that commit but skip that commit itself.

codetheweb committed:
Merge remote-tracking branch 'upstream/master' into feature/ui-touchup

codetheweb commented:
Sorry about that @Kiro47.

Looks like it's good to go now.
